apps/manager_app/get_device_info.c

```diff
@@ -0,0 +1,185 @@
+ * @brief Returns the formatted semantic versioning.
+ *
+ * @param firmware_version
+ * @return bool Status indicating the
+ */
+bool get_firmware_version(common_version_t *firmware_version);
+
+/*****************************************************************************
+ * STATIC FUNCTIONS
+ *****************************************************************************/
+
+bool get_firmware_version(common_version_t *firmware_version) {
+  uint32_t version = get_fwVer();
+
+  if (NULL == firmware_version || 0 == version || 0xFFFFFFFF == version) {
+    return false;
+  }
+
+  firmware_version->major = (version >> 24) & 0xFF;
+  firmware_version->minor = (version >> 16) & 0xFF;
+  firmware_version->patch = version & 0xFFFF;
+  return true;
+}
+
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+
+manager_get_device_info_response_t get_device_info(void) {
+  manager_get_device_info_response_t device_info =
+      MANAGER_GET_DEVICE_INFO_RESPONSE_INIT_ZERO;
+  uint32_t status = get_device_serial();
+
+  device_info.which_response = MANAGER_GET_DEVICE_INFO_RESPONSE_RESULT_TAG;
+  if (status != ATCA_SUCCESS) {
+    device_info.which_response =
+        MANAGER_GET_DEVICE_INFO_RESPONSE_CORE_ERROR_TAG;
+    LOG_CRITICAL("serial %d", status);
+  }
+
+  if (device_info.which_response ==
+      MANAGER_GET_DEVICE_INFO_RESPONSE_RESULT_TAG) {
+    manager_get_device_info_result_response_t *result =
+        &device_info.response.result;
+    result->has_firmware_version =
+        get_firmware_version(&result->firmware_version);
+    memcpy(result->device_serial, atecc_data.device_serial, DEVICE_SERIAL_SIZE);
+    result->is_authenticated = (is_device_authenticated() == 1);
+    result->is_initial = true;            // TODO: Get from memory
+    result->has_initial_states = true;    // TODO: Get from memory
+    // TODO: populate applet list (result->applet_list)
+  } else {
+    device_info.response.core_error.which_error =
+        ERROR_CORE_ERROR_UNKNOWN_ERROR_TAG;
+    device_info.response.core_error.error.unknown_error = status;
+  }
+
+  return device_info;
+}
+
+void get_device_info_flow(const manager_query_t *query) {
+  size_t msg_size = 0;
+  uint8_t response[MANAGER_GET_DEVICE_INFO_RESULT_RESPONSE_SIZE] = {0};
+  manager_result_t result = MANAGER_RESULT_INIT_ZERO;
+
+  if (MANAGER_QUERY_GET_DEVICE_INFO_TAG != query->which_request ||
+      MANAGER_GET_DEVICE_INFO_REQUEST_INITIATE_TAG !=
+          query->request.get_device_info.which_request) {
+    result.which_response = MANAGER_RESULT_GET_DEVICE_INFO_TAG;
+    result.response.get_device_info.which_response =
+        MANAGER_GET_DEVICE_INFO_RESPONSE_CORE_ERROR_TAG;
+    result.response.get_device_info.response.core_error.which_error =
+        ERROR_CORE_ERROR_UNKNOWN_ERROR_TAG;
+    result.response.get_device_info.response.core_error.error.unknown_error = 1;
```

Comment:
Seems this part is common in the above API as well, can we create a separate function for this?

Reply:
changed